gapis/api/vulkan/query_timestamps.go

```diff
@@ -0,0 +1,437 @@
+				cb,
+				out,
+				query)
+			newCmdBUffers[j*2+2] = commandbuffer
+
+			begin := &path.Command{
+				Indices: []uint64{uint64(id), uint64(i), uint64(j), 0},
+			}
+			c, ok := GetState(s).CommandBuffers().Lookup(buf)
+			if !ok {
+				fmt.Errorf("Invalid command buffer %v", buf)
+			}
+			n := c.CommandReferences().Len()
+			end := &path.Command{
+				Indices: []uint64{uint64(id), uint64(i), uint64(j), uint64(n - 1)},
+			}
+			t.timestamps = append(t.timestamps, replay.Timestamps{Begin: begin, End: end})
+		}
+
+		cmdBufferPtr := allocAndRead(newCmdBUffers).Ptr()
+		newSubmitInfos[i] = NewVkSubmitInfo(s.Arena,
+			VkStructureType_VK_STRUCTURE_TYPE_SUBMIT_INFO,
+			0,                            // pNext
+			si.WaitSemaphoreCount(),      // waitSemaphoreCount
+			NewVkSemaphoreᶜᵖ(waitSemPtr), // pWaitSemaphores
+			NewVkPipelineStageFlagsᶜᵖ(waitDstStagePtr), // pWaitDstStageMask
+			newCmdCount,                        // commandBufferCount
+			NewVkCommandBufferᶜᵖ(cmdBufferPtr), // pCommandBuffers
+			si.SignalSemaphoreCount(),          // signalSemaphoreCount
+			NewVkSemaphoreᶜᵖ(signalSemPtr),     // pSignalSemaphores
+		)
+	}
+
+	submitInfoPtr := allocAndRead(newSubmitInfos).Ptr()
+
+	newCmd := cb.VkQueueSubmit(
+		cmd.Queue(),
+		cmd.SubmitCount(),
+		submitInfoPtr,
+		cmd.Fence(),
+		VkResult_VK_SUCCESS,
+	)
+
+	for _, read := range reads {
+		newCmd.AddRead(read.Data())
+	}
+	out.MutateAndWrite(ctx, id, newCmd)
+}
+
+func (t *queryTimestamps) Transform(ctx context.Context, id api.CmdID, cmd api.Cmd, out transform.Writer) {
+	s := out.State()
+	cb := CommandBuilder{Thread: cmd.Thread(), Arena: s.Arena}
+
+	defer func() {
+		for _, d := range t.allocated {
+			d.Free()
+		}
+	}()
+
+	switch cmd := cmd.(type) {
+	case *VkGetPhysicalDeviceProperties:
+		cmd.Extras().Observations().ApplyWrites(s.Memory.ApplicationPool())
+		out.MutateAndWrite(ctx, id, cmd)
+		info := cmd.pProperties.MustRead(ctx, cmd, s, nil)
+		t.timestampPeriod = info.Limits().TimestampPeriod()
+		log.I(ctx, "TimestampPeriod is %v", t.timestampPeriod)
+
+	case *VkCreateDevice:
+		out.MutateAndWrite(ctx, id, cmd)
+		cmd.Extras().Observations().ApplyWrites(s.Memory.ApplicationPool())
+		t.device = cmd.pDevice.MustRead(ctx, cmd, s, nil)
+
+	case *VkQueueSubmit:
+		cmd.Extras().Observations().ApplyReads(s.Memory.ApplicationPool())
+		queue := cmd.Queue()
+		queueFamilyIndex := GetState(s).Queues().Get(queue).Family()
+		submitCount := cmd.SubmitCount()
+		submitInfos := cmd.pSubmits.Slice(0, uint64(submitCount), s.MemoryLayout).MustRead(ctx, cmd, s, nil)
+		cmdBufferCount := uint32(0)
+		for i := uint32(0); i < submitCount; i++ {
+			si := submitInfos[i]
+			cmdBufferCount += si.CommandBufferCount()
+		}
+		queryCount := cmdBufferCount * 2
+
+		t.createComandpoolIfNeeded(ctx, cb, out, queueFamilyIndex)
+		t.createQueryPoolIfNeeded(ctx, cb, out, queryCount)
+		t.rewriteQueueSubmit(ctx, cb, out, id, cmd)
+
+		waitCmd := cb.VkQueueWaitIdle(queue, VkResult_VK_SUCCESS)
```

Comment:
As discussed, we probably want to, in a follow-up CL, not vkQueueWaitIdle every queue submit, (which may cause stalls), we should probably accumulate results until we are done, or our query-pool is full, and then wait at that point.

Eventually it might be even better to do a pre-pass over the commands to determine what size of query pool we want, and then we can just flush at the end.

Reply:
added a TODO for tracking.